runtime/doc/develop.txt

```diff
@@ -166,77 +166,282 @@ stick to these rules, to keep the sources readable and maintainable.
-problems for existing patches.  Only use them for new and changed code.
+Features not to be used ~
 
-Comments ~
+These C99 features are not to be used, because not enough compilers support
+them:
+- Variable length arrays (even in C11 this is an optional feature).
+- C99 _Bool and _Complex types.
+- "inline" (it's hardly ever needed, let the optimizer do its work)
+- flexible array members: Not supported by HP-UX C compiler (John Marriott)
 
-Traditionally Vim uses /* comments */.  We intend to keep it that way
-for file and function headers and larger blocks of code, E.g.:
+COMMENTS					       *style-comments*
+
+Comments are good, but there is also a danger of over-commenting. NEVER try to
+explain HOW your code works in a comment. Write the code so that the working
+is obvious.
+
+Generally, you want your comments to tell WHAT your code does and WHY it does
+it, instead of HOW it is done.
+
+Try to avoid putting multiline comments inside a function body: if the
+function is so complex that you need to separately comment parts of it, you
+should probably rethink the structure of the function.
+
+Avoid using a comment to explain what a variable is, instead of naming the
+variable sensibly.
+
+For multiline comments use:
+>
+    /*
+     * comment
+     */
+<
+For inline comments use:
+>
+    int var; // comment
+<
+
+
+INDENTATION					      *style-indentation*
+
+We use 4 space to indent the code. If you are using vim, to edit the source,
+you don't need to do anything due to the |modeline|.
+
+For other editors an editorconfig is provided at the root of the repo.
+
+
+DECLARATIONS	                                     *style-declarations*
+
+Declare for loop variable in the guard (see also |style-loops|):
+Ok:
+>
+    for (int i = 0; i < len; ++i)
+<
+Wrong:
+>
+    int i;
+    for (i = 0; i < len; ++i)
+<
+Always declare a variable with a default value:
+Ok:
+>
+    int n = 0;
+    int *ptr = NULL;
+<
+Wrong:
+>
+    int n;
+    int *ptr;
+<
+
+
+LOOPS						            *style-loops*
+
+Use while loops when:
+- You need to iterate over a data structure that needs to move some sort of
+  pointer, such as a linked list.
+- You need to iterate indefinatly (aka while (TRUE)).
+
+Use for loops when:
+- You need to iterate n time, and you don't need to modify the variable that
+  index
+- There is a for each macro ready to use for your purpose
+
+Note Don't: for (;;) or for (; guard; increment)
+
+
+BRACES							   *style-braces*
+
+All curly braces must be returned onto a new line:
+OK:
+>
+    if (cond)
+    {
+       cmd;
+       cmd;
+    }
+    else
+    {
+       cmd;
+       cmd;
+    }
+<
+Wrong:
+>
+    if (cond) {
+       cmd;
+       cmd;
+    } else {
+       cmd;
+       cmd;
+    }
+<
+Ok:
+>
+    while (cond)
+    {
+	cmd;
+    }
+<
+Wrong:
+>
+    while (cond) {
+	cmd;
+    }
+<
+When a block has one line, including comments, the braces can be left out.
+OK:
+>
+    if (cond)
+	cmd;
+    else
+	cmd;
```

Comment:
Currently, we don't use braces for a one-line block. However, this can be annoying when adding or removing a line from the block (mainly when debugging).

Also, there was a famous bug because of not using braces:
https://www.imperialviolet.org/2014/02/22/applebug.html
Moreover, recent languages like Go and Rust cannot omit the braces.

Omitting the braces doesn't have benefits other than the concise looks.

@chrisbra 
How about also accepting using braces for a one-line block?